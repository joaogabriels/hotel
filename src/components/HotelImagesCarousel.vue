<template>
  <q-carousel animated v-model="currentSlide" infinite :height="height" class="hotel-carousel" ref="carouselRef"
    :padding="false">
    <q-carousel-slide v-for="(image, index) in images" :name="index" :key="index">
      <img :src="image" @error="onImageError" class="full-width full-height fit" />
    </q-carousel-slide>

    <template v-slot:control>
      <q-carousel-control position="left" class="carousel-nav-btn" :offset="[16, middleOffset]">
        <q-btn round dense color="white" text-color="black" icon="arrow_back" @click="carouselRef?.previous()"
          class="carousel-nav-btn" size="12px" />
      </q-carousel-control>

      <q-carousel-control position="right" class="carousel-nav-btn" :offset="[16, middleOffset]">
        <q-btn round dense color="white" text-color="black" icon="arrow_forward" @click="carouselRef?.next()"
          class="carousel-nav-btn" size="12px" />
      </q-carousel-control>

      <q-carousel-control v-if="showCounter" position="bottom" class="slide-counter-container">
        <div class="slide-counter">
          {{ currentSlide + 1 }}/{{ images.length }}
        </div>
      </q-carousel-control>
    </template>
  </q-carousel>

</template>

<script setup lang="ts">
import { computed, ref } from 'vue';
import type { QCarousel } from 'quasar';

const currentSlide = ref(0);
const carouselRef = ref<QCarousel>();

const props = defineProps<{
  images: string[];
  height?: string;
  arrows?: boolean;
  showCounter?: boolean;
}>();
const middleOffset = computed(() => {
  const heightNum = parseInt((props.height ?? '0').replace('px', ''));

  return heightNum / 2;
});

const onImageError = (event: Event) => {
  const img = event.target as HTMLImageElement;
  img.src = 'https://placehold.co/600x400?text=Imagem+n√£o+encontrada';
};
</script>

<style lang="scss" scoped>
.hotel-carousel {
  border-radius: inherit;
  position: relative;

  .q-carousel__slide {
    @media screen and (max-width: 768px) {
      padding: 16px;
    }
  }

  .carousel-nav-btn {
    opacity: 0.9;
    transition: opacity 0.3s ease;
    margin-top: 0;

    &:hover {
      opacity: 1;
    }
  }

  .slide-counter-container {
    display: flex;
    align-items: center;
    justify-content: center;

    .slide-counter {
      background: rgba(0, 0, 0, 0.7);
      color: white;
      border-radius: 1000px;
      font-size: 12px;
      padding: 4px 8px;
    }
  }

  img {
    object-fit: cover;
  }

  :deep(.q-carousel__slide) {
    padding: 0;
  }

  :deep(.q-panel) {
    overflow: hidden;
  }
}
</style>