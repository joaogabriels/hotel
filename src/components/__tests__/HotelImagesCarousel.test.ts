import { mount } from '@vue/test-utils'
import { describe, expect, it } from 'vitest'
import { installQuasarPlugin } from '@quasar/quasar-app-extension-testing-unit-vitest'

import HotelImagesCarousel from '../HotelImagesCarousel.vue'
import { mockHotel } from './mocks'

installQuasarPlugin()

describe('HotelImagesCarousel', () => {
  it('matches snapshot', () => {
    const wrapper = mount(HotelImagesCarousel, {
      props: {
        images: mockHotel.images,
        height: '320px',
        arrows: true,
        showCounter: false,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('matches snapshot with counter', () => {
    const wrapper = mount(HotelImagesCarousel, {
      props: {
        images: mockHotel.images,
        height: '320px',
        arrows: true,
        showCounter: true,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should trigger next button', async () => {
    const wrapper = mount(HotelImagesCarousel, {
      props: {
        images: mockHotel.images,
        height: '320px',
        arrows: true,
        showCounter: true,
      },
    })

    const nextButton = wrapper.findAll('.q-btn')[1]
    await nextButton?.trigger('click')

    expect((wrapper.vm as unknown as { currentSlide: number }).currentSlide).toBe(1)
  })

  it('should trigger back button', async () => {
    const wrapper = mount(HotelImagesCarousel, {
      props: {
        images: mockHotel.images,
        height: '320px',
        arrows: true,
        showCounter: true,
      },
    })

    const nextButton = wrapper.findAll('.q-btn')[1]
    await nextButton?.trigger('click')

    expect((wrapper.vm as unknown as { currentSlide: number }).currentSlide).toBe(1)

    const backButton = wrapper.findAll('.q-btn')[0]
    await backButton?.trigger('click')

    expect((wrapper.vm as unknown as { currentSlide: number }).currentSlide).toBe(0)
  })
})
