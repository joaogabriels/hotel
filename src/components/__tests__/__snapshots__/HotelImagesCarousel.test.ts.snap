// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HotelImagesCarousel > matches snapshot 1`] = `
"<div data-v-f3901ebe=\\"\\" class=\\"q-carousel q-panel-parent q-carousel--without-padding hotel-carousel\\" style=\\"height: 320px;\\">
  <div class=\\"q-carousel__slides-container\\">
    <transition-stub appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
      <div class=\\"q-panel scroll\\" style=\\"--q-transition-duration: 300ms;\\" role=\\"tabpanel\\">
        <div data-v-f3901ebe=\\"\\" class=\\"q-carousel__slide\\"><img data-v-f3901ebe=\\"\\" src=\\"https://fakeimg.pl/600x400\\" class=\\"full-width full-height fit\\"></div>
      </div>
    </transition-stub>
  </div>
  <div data-v-f3901ebe=\\"\\" class=\\"q-carousel__control absolute absolute-left carousel-nav-btn\\" style=\\"margin: 160px 16px;\\"><button data-v-f3901ebe=\\"\\" class=\\"q-btn q-btn-item non-selectable no-outline q-btn--standard q-btn--round bg-white text-black q-btn--actionable q-focusable q-hoverable q-btn--dense carousel-nav-btn\\" style=\\"font-size: 12px;\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"q-focus-helper\\"></span><span class=\\"q-btn__content text-center col items-center q-anchor--skip justify-center row\\"><i class=\\"q-icon notranslate material-icons\\" aria-hidden=\\"true\\" role=\\"img\\">arrow_back</i></span></button></div>
  <div data-v-f3901ebe=\\"\\" class=\\"q-carousel__control absolute absolute-right carousel-nav-btn\\" style=\\"margin: 160px 16px;\\"><button data-v-f3901ebe=\\"\\" class=\\"q-btn q-btn-item non-selectable no-outline q-btn--standard q-btn--round bg-white text-black q-btn--actionable q-focusable q-hoverable q-btn--dense carousel-nav-btn\\" style=\\"font-size: 12px;\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"q-focus-helper\\"></span><span class=\\"q-btn__content text-center col items-center q-anchor--skip justify-center row\\"><i class=\\"q-icon notranslate material-icons\\" aria-hidden=\\"true\\" role=\\"img\\">arrow_forward</i></span></button></div>
  <div data-v-f3901ebe=\\"\\" class=\\"q-carousel__control absolute absolute-bottom slide-counter-container\\" style=\\"margin: 18px;\\">
    <div data-v-f3901ebe=\\"\\" class=\\"slide-counter\\">1/3</div>
  </div>
</div>"
`;

exports[`HotelImagesCarousel > matches snapshot with counter 1`] = `
"<div data-v-f3901ebe=\\"\\" class=\\"q-carousel q-panel-parent q-carousel--without-padding hotel-carousel\\" style=\\"height: 320px;\\">
  <div class=\\"q-carousel__slides-container\\">
    <transition-stub appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
      <div class=\\"q-panel scroll\\" style=\\"--q-transition-duration: 300ms;\\" role=\\"tabpanel\\">
        <div data-v-f3901ebe=\\"\\" class=\\"q-carousel__slide\\"><img data-v-f3901ebe=\\"\\" src=\\"https://fakeimg.pl/600x400\\" class=\\"full-width full-height fit\\"></div>
      </div>
    </transition-stub>
  </div>
  <div data-v-f3901ebe=\\"\\" class=\\"q-carousel__control absolute absolute-left carousel-nav-btn\\" style=\\"margin: 160px 16px;\\"><button data-v-f3901ebe=\\"\\" class=\\"q-btn q-btn-item non-selectable no-outline q-btn--standard q-btn--round bg-white text-black q-btn--actionable q-focusable q-hoverable q-btn--dense carousel-nav-btn\\" style=\\"font-size: 12px;\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"q-focus-helper\\"></span><span class=\\"q-btn__content text-center col items-center q-anchor--skip justify-center row\\"><i class=\\"q-icon notranslate material-icons\\" aria-hidden=\\"true\\" role=\\"img\\">arrow_back</i></span></button></div>
  <div data-v-f3901ebe=\\"\\" class=\\"q-carousel__control absolute absolute-right carousel-nav-btn\\" style=\\"margin: 160px 16px;\\"><button data-v-f3901ebe=\\"\\" class=\\"q-btn q-btn-item non-selectable no-outline q-btn--standard q-btn--round bg-white text-black q-btn--actionable q-focusable q-hoverable q-btn--dense carousel-nav-btn\\" style=\\"font-size: 12px;\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"q-focus-helper\\"></span><span class=\\"q-btn__content text-center col items-center q-anchor--skip justify-center row\\"><i class=\\"q-icon notranslate material-icons\\" aria-hidden=\\"true\\" role=\\"img\\">arrow_forward</i></span></button></div>
  <div data-v-f3901ebe=\\"\\" class=\\"q-carousel__control absolute absolute-bottom slide-counter-container\\" style=\\"margin: 18px;\\">
    <div data-v-f3901ebe=\\"\\" class=\\"slide-counter\\">1/3</div>
  </div>
</div>"
`;
